src/java.base/share/classes/jdk/internal/util/ClassFileDumper.java

```diff
@@ -0,0 +1,234 @@
+
+    private ClassFileDumper(String key, Path path) {
+        this.key = key;
+        this.dumpDir = path;
+    }
+
+    public String key() {
+        return key;
+    }
+    public boolean isEnabled() {
+        return dumpDir != null;
+    }
+
+    public Path dumpPath() {
+        return dumpDir;
+    }
+
+    public Path pathname(String internalName) {
+        return dumpDir.resolve(encodeForFilename(internalName) + ".class");
+    }
+
+    /**
+     * This method determines the path name from the given name and {@code Class}
+     * object.  If it is a hidden class, it will dump the given bytes at
+     * a path of the given name with a suffix "." concatenated
+     * with the suffix of the hidden class name.
+     */
+    public void dumpClass(String name, Class<?> c, byte[] bytes) {
+        if (!isEnabled()) return;
+
+        String cn = c.getName();
+        int suffixIdx = cn.lastIndexOf('/');
+        if (suffixIdx > 0) {
+            name += '.' + cn.substring(suffixIdx + 1);
+        }
+        write(pathname(name), bytes);
+    }
+
+    /**
+     * This method dumps the given bytes at a path of the given name with
+     * a suffix ".failed-$COUNTER" where $COUNTER will be incremented
+     * for each time this method is called.
+     */
+    public void dumpFailedClass(String name, byte[] bytes) {
+        if (!isEnabled()) return;
+
+        write(pathname(name + ".failed-" + counter.incrementAndGet()), bytes);
+    }
+
+    @SuppressWarnings("removal")
+    private void write(Path path, byte[] bytes) {
+        AccessController.doPrivileged(new PrivilegedAction<>() {
+                @Override public Void run() {
+                    try {
+                        Path dir = path.getParent();
+                        Files.createDirectories(dir);
+                        Files.write(path, bytes);
+                    } catch (Exception ex) {
+                        if (VM.isModuleSystemInited()) {
+                            // log only when lambda is ready to use
+                            System.getLogger(ClassFileDumper.class.getName())
+                                  .log(System.Logger.Level.WARNING, "Exception writing to " +
+                                          path.toString() + " " + ex.getMessage());
+                        }
+                        // simply don't care if this operation failed
+                    }
+                    return null;
+                }},
+                null,
+                new FilePermission("<<ALL FILES>>", "read, write"),
+                // createDirectories may need it
+                new PropertyPermission("user.dir", "read"));
+    }
+
+    @SuppressWarnings("removal")
+    private static Path validateDumpDir(Path path) {
+            return AccessController.doPrivileged(new PrivilegedAction<>() {
+                @Override
+                public Path run() {
+                    if (!Files.exists(path)) {
```

Comment:
I do not think this is necessary, Files.createDirectories() already only create directories if they do not exist.
Moreover, someone can change the state of the files in between the two calls (Files.exists and Files.createDirectories) given that those two calls are not atomic from the filesystem POV.

Reply:
that's right. updated